python_modules/dagster-ext/dagster_ext/__init__.py

```diff
@@ -543,6 +543,80 @@ def upload_messages_chunk(self, payload: IO, index: int) -> None:
 # ########################
 
 
+class IContext(ABC):
+    """Base class for asset context implemented by AssetExecutionContext and ExtContext."""
+
+    @property
+    @abstractmethod
+    def is_asset_step(self) -> bool:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def asset_key(self) -> str:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def asset_keys(self) -> Sequence[str]:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def provenance(self) -> Optional[ExtDataProvenance]:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def provenance_by_asset_key(self) -> Mapping[str, Optional[ExtDataProvenance]]:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def code_version(self) -> Optional[str]:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def code_version_by_asset_key(self) -> Mapping[str, Optional[str]]:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def is_partition_step(self) -> bool:
```

Comment:
the equivalent method in `OpExecutionContext` is `has_partition_key`. We should align on the name. Options:
* `is_partition_step`
* `has_partition_key`
* `is_partitioned`
* ...

Reply:
I like `is_partitioned_step` or `is_partitioned_execution` with slight preference for `is_partitioned_execution`. Some thought process/reasoning:
* `is_partitioned_step` adds the concept of a step. It's not hard to understand what a step is, but it's still one additional thing. 
* `is_partitioned_execution` uses a more familiar term, but is there a possibility that an asset could not be partitioned but still be part of a partitioned run? thinking of an asset with an `AllPartitionMapping`. I will test this out.

Reply:
i like `is_partitioned`. @smackesey  what are our guarantees about the stability of ExtContext?  if i rename `is_partitioned_step` to `is_partitioned` do i need to add a deprecation warning to `is_partitioned_step` or can i just delete it?

Reply:
Our guarantees are weak right now, I think it's fine to change anything on there in the next few days.

That said I don't really like `is_partitioned`, because we have a lot of concepts floating around and it's best to be clear about what applies to what. `is_partitioned` makes it sound like the context itself is partitioned.

I think `is_partitioned_execution` could be good _if_ we use that terminology elsewhere. `is_partitioned_step` also, though I think it wasn't chosen in a very principled way. 

Curious what @sryza and @schrockn think here-- I believe in our ontology, jobs and assets are two entities that can clearly be the subject of partitioning. How do we refer to a step corresponding to one or more partitions?

Reply:
On the asset execution context we have `has_partition_key` and `has_partition_key_range`. Is it functionally different than those methods on the asset execution context?  My instinct would be to be consistent, even if we don't think it's the optimal choice.

Reply:
Agree with @smackesey  that `is_partitioned` is not specific enough.

I concur with @sryza. Having `has_partition_key` and `has_partition_key_range` be two sep methods here is desirable because it accurately communicates that there are two different APIs for accessing partitioning information. If it is _only_ `is_partitioned_step`, the user will sometimes need to make a further distinction between `has_partition_key` and `has_partition_key_range` anyways. I do prefer `has_single_partition_key` to `has_partition_key` as someone reasonable might think that `has_partition_key` should return True whenever `has_partition_key_range` is True.

I think the crux of this is this image:

![image](https://github.com/dagster-io/dagster/assets/28738937/5a11ec93-3799-4a02-bf95-87b6a2719d4d)

What is the behavior and best naming in the "potentially confusing" quadrant of the state space above.